src/java.base/share/classes/jdk/internal/classfile/constantpool/ConstantPoolBuilder.java

```diff
@@ -0,0 +1,574 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.classfile.constantpool;
+
+import java.lang.constant.ClassDesc;
+import java.lang.constant.ConstantDesc;
+import java.lang.constant.DirectMethodHandleDesc;
+import java.lang.constant.DynamicCallSiteDesc;
+import java.lang.constant.DynamicConstantDesc;
+import java.lang.constant.MethodTypeDesc;
+import java.util.Collection;
+import java.util.List;
+
+import jdk.internal.classfile.BootstrapMethodEntry;
+import jdk.internal.classfile.BufWriter;
+import jdk.internal.classfile.ClassBuilder;
+import jdk.internal.classfile.ClassModel;
+import jdk.internal.classfile.ClassReader;
+import jdk.internal.classfile.Classfile;
+import jdk.internal.classfile.impl.ClassReaderImpl;
+import jdk.internal.classfile.impl.Options;
+import jdk.internal.classfile.jdktypes.ModuleDesc;
+import jdk.internal.classfile.jdktypes.PackageDesc;
+import jdk.internal.classfile.WritableElement;
+import jdk.internal.classfile.impl.SplitConstantPool;
+import jdk.internal.classfile.impl.TemporaryConstantPool;
+import jdk.internal.classfile.impl.Util;
+
+/**
+ * Builder for the constant pool of a classfile.  Provides read and write access
+ * to the constant pool that is being built.  Writing is append-only and idempotent
+ * (entry-bearing methods will return an existing entry if there is one).
+ *
+ * A {@linkplain ConstantPoolBuilder} is associated with a {@link ClassBuilder}.
+ * The {@linkplain ConstantPoolBuilder} also provides access to some of the
+ * state of the {@linkplain ClassBuilder}, such as classfile processing options.
+ */
+public sealed interface ConstantPoolBuilder
+        extends ConstantPool, WritableElement<ConstantPool>
+        permits SplitConstantPool, TemporaryConstantPool {
+
+    /**
+     * {@return a new constant pool builder}  The new constant pool builder
+     * will inherit the classfile processing options of the specified class.
+     * If the processing options include {@link Classfile.Option.Key#CP_SHARING},
+     * (the default) the new constant pool builder will be also be pre-populated with the
+     * contents of the constant pool associated with the class reader.
+     *
+     * @param classModel the class to copy from
+     */
+    static ConstantPoolBuilder of(ClassModel classModel) {
+        ClassReader reader = (ClassReader) classModel.constantPool();
+        return reader.optionValue(Classfile.Option.Key.CP_SHARING)
```

Comment:
If you propose `constantPoolSharing(false)` should automatically imply `processUnknownAttributes(false)` - it would avoid potential invalid CP entries in the unknown attributes and we may consider that dependency between the two options.
However on the other side there might be unknown attributes completely independent of CP and for such the transformation is safe. Strong options implication would force users to create custom attribute mappers for such attributes.
Current options independence gives user power to configure transformations flexibly, however on the other side it allows to produce classes with invalid CP entries in unknown attributes.